### PR TITLE
fix(evm): enforce EIP-7825 per-tx gas cap once Osaka activates

### PR DESCRIPTION
**Pull request by JayT106**
## Summary

- Enforce EIP-7825's per-tx gas cap (geth's hardcoded `MaxTxGas` constant) once Osaka activates, in both `ApplyMessageWithConfig` (covers `FinalizeBlock`, `eth_call`, `eth_estimateGas`) and the ante handler's `CheckEthGasConsume`, so oversized txs get rejected before entering the mempool, not just at execution.
- Also enforce it in `applyCall` (the `eth_simulateV1` path), keeping it in sync with `ApplyMessageWithConfig`.
- `eth_call`/`eth_estimateGas`/`eth_simulateV1` messages (`SkipTransactionChecks`) still bypass the cap.
- Extend tests to cover the ante-handler rejection, ReCheckTx cap rejection, and the pre-Osaka no-op case.

### CODE REVIEW

**Comment by JayT106** [comment]
@claude review

**Review by XinyuCRO — COMMENTED** [review]
need also add ante level check to prevent tx included in mempool based on https://eips.ethereum.org/EIPS/eip-7825

File: x/evm/keeper/state_transition.go [hunk 1]
```diff
@@ -423,6 +423,13 @@ func (k *Keeper) ApplyMessageWithConfig(
 	}
 
 	rules := cfg.Rules
+
+	// Enforce EIP-7825 per-tx gas cap for Osaka in all execution contexts,
+	// including eth_call and FinalizeBlock paths that bypass the ante handler.
+	if !msg.SkipTransactionChecks && rules.IsOsaka && msg.GasLimit > params.MaxTxGas {
```
Comment (XinyuCRO) [inline x/evm/keeper/state_transition.go]: user might have been charged full gasLimit, the geth rejects whole block if such tx happens here, don't know if its easy to achieve

Comment (thomas-nguy) [inline x/evm/keeper/state_transition.go]: nitpick, but could refactor it into `precheck` to follow ethereum architecture

**Comment by thomas-nguy** [comment]
Hi @JayT106 

Does this PR is to fix a vulnerability or ONLY to follow Ethereum standard?

Since our chain is cosmos based, we cannot follow completely the Ethereum standard but we do have some protection such as the ante handler preventing the transaction gas to exceed the block gas.

In case we want to implement a transaction gas cap. Should we introduce a new consensus param into the evm module instead of the geth hard coded one?

**Comment by JayT106** [comment]
> LY to follow Ether

 To follow Ethereum standard. But may still worth to have it for the security consideration and the Rationale in 
https://eips.ethereum.org/EIPS/eip-7825

> hould we introduce a new consensus param into the evm module instead of the geth hard coded one?

Indeed the a new consensus param will be better since cronos have different block gas limit. Do you want to proceed it in this way?

File: x/evm/keeper/state_transition.go (hunk 1 above)
Comment (JayT106) [reply x/evm/keeper/state_transition.go]: It doesn't charge full gasLimit like geth, the flow:
```
 ApplyMessageWithConfig error → ApplyTransaction returns err → EthereumTx msg handler returns err → whole cosmos tx fails, entire runTx cache rolled
  back (ante handler's fee deduction included)
```

Comment (JayT106) [reply x/evm/keeper/state_transition.go]: done!

**Comment by JayT106** [comment]
> 



> need also add ante level check to prevent tx included in mempool based on https://eips.ethereum.org/EIPS/eip-7825

added check in https://github.com/crypto-org-chain/ethermint/pull/1056/commits/c9dec389dbafe41720e39ccac42f0d4e963c9602

**Comment by thomas-nguy** [comment]
`Indeed the a new consensus param will be better since cronos have different block gas limit. Do you want to proceed it in this way?`

Sure

**Review by thomas-nguy — APPROVED** [review]

**Comment by thomas-nguy** [comment]
@JayT106 please run an analysis to see if we are not missing any corner cases

**Comment by JayT106** [comment]
@claude review

File: x/evm/keeper/state_transition.go [hunk 2]
```diff
@@ -423,6 +440,12 @@ func (k *Keeper) ApplyMessageWithConfig(
 	}
 
 	rules := cfg.Rules
+
+	// Duplicated here since eth_call/FinalizeBlock bypass the ante handler.
+	if err := preCheckMaxTxGas(msg, rules); err != nil {
```
[the remark this answers is not in the document]
Comment (JayT106) [reply x/evm/keeper/state_transition.go:445]: Fixed: added `ErrorIs` sentinel check (`core.ErrGasLimitTooHigh`) alongside the message-substring check for all three MaxTxGas cap ante test cases, and documented the debug_traceTransaction replay edge case on `preCheckMaxTxGas`.

File: evmd/ante/eth_test.go [hunk 3]
```diff
@@ -438,13 +450,13 @@ func (suite *AnteTestSuite) TestEthGasConsumeDecorator() {
...
+			overMaxTxGasTx,
 			math.MaxUint64,
 			func() {},
 			false, false,
 			0,
-			nil,
+			fmt.Errorf("cap: %d, tx: %d", params.MaxTxGas, overMaxTxGasLimit),
 		},
 		{
 			"not enough balance for fees",
```
[the remark this answers is not in the document]
Comment (JayT106) [reply evmd/ante/eth_test.go:467]: Fixed: added `ErrorIs` sentinel check alongside the message-substring check for all three MaxTxGas cap ante test cases.

**Comment by JayT106** [comment]
@claude review
